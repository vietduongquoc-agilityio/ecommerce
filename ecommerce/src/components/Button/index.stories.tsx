import type { Meta, StoryObj } from "@storybook/react";
import Button from "./";

const meta: Meta<typeof Button> = {
  title: "Components/Button",
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    onClick: { action: "clicked" },
    variant: {
      control: "select",
      options: [
        "primary",
        "secondary",
        "outline",
        "ghost",
        "disabled",
        "primaryXs",
      ],
    },
    size: {
      control: "select",
      options: ["sm", "md", "lg"],
    },
    fullWidth: { control: "boolean" },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const Primary: Story = {
  args: {
    children: "Primary Button",
    variant: "primary",
    size: "md",
  },
};

export const Secondary: Story = {
  args: {
    children: "Secondary Button",
    variant: "secondary",
  },
};

export const Outline: Story = {
  args: {
    children: "Outline Button",
    variant: "outline",
  },
};

export const Disabled: Story = {
  args: {
    children: "Disabled Button",
    variant: "disabled",
  },
};

export const FullWidth: Story = {
  args: {
    children: "Full Width Button",
    variant: "primary",
    fullWidth: true,
  },
};
