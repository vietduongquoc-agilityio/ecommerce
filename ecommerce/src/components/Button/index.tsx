"use client";

import { useState } from "react";
import { colors, fonts } from "@/themes";

interface ButtonProps {
  children: React.ReactNode;
  onClick?: () => void;
  variant?:
    | "primary"
    | "secondary"
    | "outline"
    | "ghost"
    | "disabled"
    | "primaryXs";
  size?: "sm" | "md" | "lg";
  fullWidth?: boolean;
  fullHeight?: boolean;
}

const Button = ({
  children,
  onClick,
  variant = "primary",
  size = "sm",
  fullWidth = false,
  fullHeight = false,
}: ButtonProps) => {
  const [isHovered, setIsHovered] = useState(false);

  const baseStyle: React.CSSProperties = {
    padding:
      size === "sm" ? "12px 52px" : size === "lg" ? "14px 24px" : "12px 36px",
    fontSize:
      size === "sm"
        ? fonts.size.sm
        : size === "lg"
        ? fonts.size.lg
        : fonts.size.md,
    fontWeight: fonts.weight.semiBold,
    textAlign: "center",
    width: fullWidth ? "100%" : "auto",
    height: fullHeight ? "100%" : "auto",
    cursor: variant === "disabled" ? "not-allowed" : "pointer",
    transition: "all 0.3s ease",
    border: "none",
  };

  const variantStyles: Record<string, React.CSSProperties> = {
    primary: {
      backgroundColor: isHovered ? "#E6B800" : colors.primary.yellow,
      color: colors.white,
    },
    primaryXs: {
      backgroundColor: isHovered ? "#E6B800" : colors.primary.yellow,
      color: colors.white,
      fontSize: fonts.size.xs,
    },
    secondary: {
      backgroundColor: isHovered ? "#FFF5CC" : colors.white,
      color: colors.primary.yellow,
    },
    outline: {
      backgroundColor: isHovered ? colors.primary.yellow : "transparent",
      border: `2px solid ${colors.primary.yellow}`,
      color: isHovered ? colors.white : colors.primary.yellow,
    },
    ghost: {
      backgroundColor: "transparent",
      color: colors.secondary.black,
      cursor: "not-allowed",
      fontSize: fonts.size.xs,
      padding: "1px",
      fontWeight: fonts.weight.medium,
    },
    disabled: {
      backgroundColor: colors.secondary.splight,
      color: colors.white,
      opacity: 0.6,
    },
  };

  return (
    <button
      onClick={variant !== "disabled" ? onClick : undefined}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
      style={{ ...baseStyle, ...variantStyles[variant] }}
    >
      {children}
    </button>
  );
};

export default Button;
