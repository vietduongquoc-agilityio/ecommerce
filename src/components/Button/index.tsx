"use client";

import { useState } from "react";
import { colors, fonts } from "@/themes";

interface ButtonProps {
  children: React.ReactNode;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
  variant?:
    | "primary"
    | "secondary"
    | "outline"
    | "ghost"
    | "disabled"
    | "primaryXs"
    | "pagination";
  size?: "sm" | "md" | "lg";
  fullWidth?: boolean;
  fullHeight?: boolean;
  style?: React.CSSProperties;
}

const Button = ({
  children,
  onClick,
  variant = "primary",
  size = "sm",
  fullWidth = false,
  fullHeight = false,
  style = {},
}: ButtonProps) => {
  const [isHovered, setIsHovered] = useState(false);

  const baseStyle: React.CSSProperties = {
    padding:
      size === "sm" ? "12px 52px" : size === "lg" ? "14px 24px" : "12px 36px",
    fontSize:
      size === "sm"
        ? fonts.size.sm
        : size === "lg"
        ? fonts.size.lg
        : fonts.size.md,
    fontWeight: fonts.weight.semiBold,
    textAlign: "center",
    width: fullWidth ? "100%" : "auto",
    height: fullHeight ? "100%" : "auto",
    cursor: variant === "disabled" ? "not-allowed" : "pointer",
    transition: "all 0.3s ease",
    border: "none",
  };

  const variantStyles: Record<string, React.CSSProperties> = {
    primary: {
      backgroundColor: isHovered ? "#E6B800" : colors.primary.yellow,
      color: colors.white,
    },
    primaryXs: {
      backgroundColor: isHovered ? "#E6B800" : colors.primary.yellow,
      color: colors.white,
      fontSize: fonts.size.xs,
    },
    secondary: {
      backgroundColor: isHovered ? "#FFF5CC" : colors.white,
      color: colors.primary.yellow,
    },
    outline: {
      backgroundColor: isHovered ? colors.white : "transparent",
      border: isHovered ? `1px solid ${colors.secondary.splight}` :`1px solid ${colors.black}`,
      color: isHovered ? colors.secondary.splight : colors.black,
    },
    ghost: {
      backgroundColor: "transparent",
      color: colors.secondary.black,
      fontSize: fonts.size.xs,
      padding: "1px",
      fontWeight: fonts.weight.medium,
    },
    disabled: {
      backgroundColor: colors.secondary.splight,
      color: colors.white,
      opacity: 0.6,
    },
    pagination: {
      padding: "20px 25px",
      borderRadius: "10px",
      backgroundColor: isHovered ? "#B88E2F" : colors.primary.dark,
      color: isHovered ? "#FFFFFF" : colors.secondary.black,
    },
  };

  return (
    <button
      onClick={variant !== "disabled" ? onClick : undefined}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
      style={{ ...baseStyle, ...variantStyles[variant], ...style }}
    >
      {children}
    </button>
  );
};

export default Button;
