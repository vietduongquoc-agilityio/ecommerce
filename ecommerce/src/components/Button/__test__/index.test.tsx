import React from "react";
import { render } from "@testing-library/react";
import Button from "../index";

describe("Button component", () => {
  it("matches snapshot", () => {
    const { asFragment } = render(<Button variant="primary">Click me</Button>);
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders with correct text", () => {
    const { getByText } = render(<Button variant="primary">Click me</Button>);
    expect(getByText("Click me")).toBeInTheDocument();
  });

  it("renders disabled button correctly", () => {
    const { getByText } = render(<Button variant="disabled">Disabled</Button>);
    expect(getByText("Disabled")).toBeInTheDocument();
    expect(getByText("Disabled")).toHaveStyle("cursor: not-allowed");
  });
});
