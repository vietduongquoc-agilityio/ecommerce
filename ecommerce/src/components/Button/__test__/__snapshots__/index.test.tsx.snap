// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component matches snapshot 1`] = `
<DocumentFragment>
  <button
    style="padding: 12px 52px; font-size: 18px; font-weight: 600; text-align: center; width: auto; cursor: pointer; transition: all 0.3s ease; background-color: rgb(184, 142, 47); color: rgb(255, 255, 255);"
  >
    Click me
  </button>
</DocumentFragment>
`;
